subsys/bluetooth/services/bas/bas_bls.c

```diff
@@ -0,0 +1,236 @@
+ * - Bits 9–11: Charging Type
+ * - Bits 12–14: Charging Fault Reason
+ * - Bit 15: RFU
+ *
+ * For detailed specification, refer to:
+ * https://bitbucket.org/bluetooth-SIG/public/src/main/gss/
+ * org.bluetooth.characteristic.battery_level_status.yaml
+ */
+
+#define BATTERY_SHIFT              0
+#define WIRED_POWER_SHIFT          1
+#define WIRELESS_POWER_SHIFT       3
+#define BATTERY_CHARGE_STATE_SHIFT 5
+#define BATTERY_CHARGE_LEVEL_SHIFT 7
+#define BATTERY_CHARGE_TYPE_SHIFT  9
+#define CHARGING_FAULT_SHIFT       12
+
+#define BATTERY_MASK              (BIT_MASK(1) << BATTERY_SHIFT)
+#define WIRED_POWER_MASK          (BIT_MASK(2) << WIRED_POWER_SHIFT)
+#define WIRELESS_POWER_MASK       (BIT_MASK(2) << WIRELESS_POWER_SHIFT)
+#define BATTERY_CHARGE_STATE_MASK (BIT_MASK(2) << BATTERY_CHARGE_STATE_SHIFT)
+#define BATTERY_CHARGE_LEVEL_MASK (BIT_MASK(2) << BATTERY_CHARGE_LEVEL_SHIFT)
+#define BATTERY_CHARGE_TYPE_MASK  (BIT_MASK(3) << BATTERY_CHARGE_TYPE_SHIFT)
+#define CHARGING_FAULT_MASK       (BIT_MASK(3) << CHARGING_FAULT_SHIFT)
+
+/*
+ * Bitfield structure: Additional Status
+ *
+ * - Bits 0–1: Service Required
+ * - Bit 2: Battery Fault
+ * - Bits 3–7: Reserved
+ */
+#define SERVICE_REQUIRED_SHIFT 0
+#define BATTERY_FAULT_SHIFT    2
+
+#define SERVICE_REQUIRED_MASK (BIT_MASK(2) << SERVICE_REQUIRED_SHIFT)
+#define BATTERY_FAULT_MASK    (BIT_MASK(1) << BATTERY_FAULT_SHIFT)
+
+void bt_bas_bls_init(void)
+{
+	LOG_INF("Initialise BAS Battery Level Status Module");
+
+	bls.flags = 0;
+	bls.power_state = 0;
+
+	if (IS_ENABLED(CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT)) {
+		/* Set identifier flag */
+		bls.flags |= BT_BAS_BLS_FLAG_IDENTIFIER_PRESENT;
+		bls.identifier = 0;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_BAS_BLS_BATTERY_LEVEL_PRESENT)) {
+		/* Set battery level flag */
+		bls.flags |= BT_BAS_BLS_FLAG_BATTERY_LEVEL_PRESENT;
+		bls.battery_level = bt_bas_get_battery_level();
+	}
+
+	if (IS_ENABLED(CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT)) {
+		/* Set additional status flag */
+		bls.flags |= BT_BAS_BLS_FLAG_ADDITIONAL_STATUS_PRESENT;
+		bls.additional_status = 0;
+	}
+}
+
+static void indicate_cb(struct bt_conn *conn, struct bt_gatt_indicate_params *params, uint8_t err)
+{
+	if (err != 0) {
+		LOG_INF("Indication failed with error %d\n", err);
+	} else {
+		LOG_INF("Indication sent successfully\n");
+	}
+}
+
+static void bt_bas_bls_update_battery_level_status(void)
+{
+	int err;
+	const struct bt_gatt_attr *attr = bt_bas_get_bas_attr(BT_BAS_IDX_BATT_LVL_STATUS_CHAR_VAL);
+
+	if (attr) {
+		const struct bt_bas_bls le_battery_level_status = {
+			.flags = bls.flags,
+			.power_state = sys_cpu_to_le16(bls.power_state),
+#if defined(CONFIG_BT_BAS_BLS_IDENTIFIER_PRESENT)
+			.identifier = sys_cpu_to_le16(bls.identifier),
+#endif
+#if defined(CONFIG_BT_BAS_BLS_BATTERY_LEVEL_PRESENT)
+			.battery_level = bls.battery_level,
+#endif
+#if defined(CONFIG_BT_BAS_BLS_ADDITIONAL_STATUS_PRESENT)
+			.additional_status = bls.additional_status,
+#endif
+		};
```

Comment:
Did you consider actually just storing `bls` as little-endian? Not saying that it needs to be changed, just wondering if it was considering and if it would be a good idea, given that the struct is already `__packed`

Reply:
I felt it is bit risky considering the need for bit manipulation in power state field